api/v1/db/json/notes-dao.js

```diff
@@ -0,0 +1,93 @@
+const appRoot = require('app-root-path');
+const _ = require('lodash');
+const fs = require('fs');
+const config = require('config');
+
+const { serializeNotes } = require('../../serializers/notes-serializer');
+
+/**
+ * @summary Return a list of notes filtered/sorted by query parameters
+ * @function
+ * @param {Object} query Query parameters
+ * @returns {Promise} Promise object represents a list of notes
+ */
+const getNotes = query => new Promise((resolve, reject) => {
+  try {
+    const { dbDirectoryPath } = config.api;
+    if (!fs.existsSync(`${appRoot}${dbDirectoryPath}`)) {
+      reject(new Error(`DB directory path: '${dbDirectoryPath}' is invalid`));
+    }
+
+    const {
+      studentID, q, sortKey, sources, contextTypes,
+    } = query;
+    const dirPath = `${dbDirectoryPath}/${studentID}`;
+    let files;
+    let rawNotes = [];
+    try {
+      files = fs.readdirSync(`${appRoot}${dirPath}`);
+    } catch (ignore) {
+      // rawNotes should remain an empty array if directory does not exist
+    }
+
+    _.forEach(files, (file) => {
+      rawNotes.push(appRoot.require(`${dirPath}/${file}`));
+    });
+
+    if (contextTypes) {
+      const contexts = contextTypes.toString().split(',');
+      rawNotes = _.filter(rawNotes, it => _.includes(contexts, it.context.contextType));
+    }
+
+    if (q) {
+      rawNotes = _.filter(rawNotes, it => _.includes(it.note, q));
+    }
+
+    // sort first by sortKey, and then by lastModified within each sorted group
+    rawNotes = _.orderBy(
+      rawNotes,
+      [
+        sortKey.toString() === 'contextType' ? it => it.context[sortKey] : it => it[sortKey],
```

Comment:
Will `sortKey` have any chances to be parsed as other types instead of a string? Also, if I sent a request to ` https://localhost:8080/api/v1/notes?studentID=111111111&sortKey=`, I got this error:

```json
{
    "errors": [
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": "sortKey must be one of ['lastModified', 'source', 'permissions', 'contextType']",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        }
    ]
}
```

IMO, `sortKey=` should be treated as not providing any `sortKey` and since we also define a `default` value for `sortKey` in OpenAPI specification, we should treat `sortKey=` as `sortKey=<default value>`. What do you think?

Reply:
Also, do you think the logic to handle this should go into the skeleton as well?

Reply:
Since the query parameter is being correctly parsed as a string, I'll remove the usage of `.toString()` since it's unnecessary. From our discussion, we'll have to bring up how we treat `?queryParam=` and `?queryParam` in our apis in a future meeting